Parser/processing logic for diagnostic ``.plist`` files

Discookie opened the pull request:
Add parser logic for ``.plist`` files, as well as a sidebar tab for displaying them.

* Only the .plist files relevant to the open editors are loaded, using the [fast-plist](https://github.com/microsoft/node-fast-plist) dependency.
* The processor supports 2 types of reload:
  * On a 'force-reload' (eg. metadata/diagnostics changed), all relevant .plist files are read from the disk 
  * On a 'normal-reload' (eg. open editors changed), only the newly-needed .plist files are read.
* The data is then made available through functions ``getFileDiagnostics`` and ``getMultipleFileDiagnostics``.
  * The latter eliminates duplicates in the returned files, since a single files' related bugs can also be located in headers, etc.

The Issues tab shows the list of issues found in the current file, with quick-buttons to go to each bug / reproduction step.

The 'open-editors' list currently only contains the visible text editors, but not the background tabs.
It would require tracking all tab change events in the diagnostics processor, while the visible text editor changes have a built-in event.

This PR will also contain tests for the parser, as well as visualizations in the editor.
Automatic reloading of diagnostic files will not be part of this PR.
SARIF support will be added by a different PR, and will require a different parsing method.

csordasmarton commented on src/backend/parser/diagnostics.ts:
```diff
@@ -0,0 +1,20 @@
+import { DiagnosticFile } from '../types';
+
+export async function parseDiagnostics(path: string): Promise<DiagnosticFile> {
+
+    const rawFileContents = await workspace.fs.readFile(Uri.file(path));
+
+    const fileContents = new TextDecoder('utf-8').decode(rawFileContents);
+
+    const parsedContents = FastPlist.parse(fileContents);
```
I think there is no benefit to add so many new lines here:
```suggestion
    const rawFileContents = await workspace.fs.readFile(Uri.file(path));
    const fileContents = new TextDecoder('utf-8').decode(rawFileContents);
    const parsedContents = FastPlist.parse(fileContents);
```

Discookie replied:
Fixed, thanks.

Discookie commented on src/sidebar/views/issues.ts:
```diff
@@ -0,0 +1,180 @@
+                this.currentBugList![element.bugIndex].path
+                    .filter(pathElem => pathElem.kind === AnalysisPathKind.Event).length,
+                (idx) => {
+                    return { ...element, reprStep: idx };
```
Reworked the section, so the line numbers are displayed, bugs are separated by file, and the repr. path displays depth.

![image](https://user-images.githubusercontent.com/16914176/127376136-3dbc9a0a-f7c7-4c0b-a0b9-97b6ae07ee3b.png)

I'm not sure about the order of the items displayed, those can be played around with later.

Discookie commented on package.json:
```diff
@@ -57,6 +57,13 @@
           "contextualTitle": "CodeChecker"
+        },
+        {
+          "id": "codechecker.views.issues",
```
Renamed everything except `DiagnosticsAPI` - in the .plist files they provide entries named ``diagnostics``, so the name felt fitting there.

Discookie commented on src/sidebar/views/issues.ts:
```diff
@@ -0,0 +1,180 @@
+    commands,
+    window
+} from 'vscode';
+import { ExtensionApi } from '../../backend/api';
```
Thanks, I'll look into that. The only place it could really be used is in tests, and maybe for the backend.

Discookie commented:
Updated with diagnostics displayed in the editor, and reworked the `Reports in current file` section. The wording is pretty much not final, but code-wise the only part left of the PR is navigation via commands.

Discookie commented on src/sidebar/views/issues.ts:
```diff
@@ -0,0 +1,180 @@
+    command?: Command;
+}
+
+export class IssuesView implements TreeDataProvider<IssueMetadata> {
```
The Overview will also contain quick buttons to Re-analyze the current file, and such.
I think for those buttons, it should be at the top.

There's also the option to make the quick buttons physical buttons, like in GitLens, but I haven't experimented with it yet:
![image](https://user-images.githubusercontent.com/16914176/127377417-15425d82-76b0-4ac1-9380-fcf011c6fce5.png)

csordasmarton commented on src/backend/processor/diagnostics.ts:
```diff
@@ -0,0 +1,196 @@
+
+    /** Prefer calling ``reloadDiagnosticsAsync`` in an async context */
+    reloadDiagnostics(forceReload?: boolean) {
+        this.reloadDiagnosticsAsync(forceReload)
```
Do not we need to use `await` here? So this function will return with a promise too. Because in this case if I do something right after the reloadDiagnostics function call, it will not wait while the reloadDiagnostics fully finishes:
```js
reloadDiagnostics();
console.log("It will be executed first.");
```
With the await we will be able to do this:
```js
reloadDiagnostics().then(() => {
  console.log("It will be executed first.");
});
``

Discookie replied:
You're right, the concept of the function was useless, as an user could ``reloadDiagnosticsAsync()`` instead in a non-async environment.

Replaced it with a ``reloadDiagnosticsSafe()``, which handles the rejection of ``reloadDiagnostics`` with an error popup to the user. (Is there a naming convention for this type of functions? I'm not sure about the name)

csordasmarton commented on src/backend/processor/diagnostics.ts:
```diff
@@ -0,0 +1,196 @@
+
+                default:
+                    console.error(err);
+                    plistErrors = true;
```
I think we can remove the `plistErrors` variable because we set it to true only here and later we print an error message based on this variable:
```suggestion
                    window.showErrorMessage('Failed to read some CodeChecker diagnostic data\nCheck console for more details');
```

Discookie replied:
The plistErrors flag is used to prevent showing multiple motifications of the same type, for example when 2 diagnostic files are corrupted.
In that case, both errors' stack traces are logged to the console, but only one notification is shown to the user inside Code.

csordasmarton commented on src/backend/processor/diagnostics.ts:
```diff
@@ -0,0 +1,196 @@
+            }
+        }
+
+        this._selectedEntry = undefined;
```
Why don't we assign `null` to this variable instead of `undefined`. `undefined` means that the variable is declared but no value is assigned to it.
```suggestion
        this._selectedEntry = null;
```
Also we can initialize this variable to null to identify that no entry is selected.

Discookie replied:
In Typescript, using `undefined` is just more convenient in variables (eg. with the ``variable?: Type`` syntax for declaration.) With null, ``Type | null`` has to be explicitly written out in every type definition.

I did use it as a return value though, as there's no syntax for an optional return.

csordasmarton commented on src/backend/types/internal/metadata.ts:
```diff
@@ -3,29 +3,29 @@
     export interface AnalyzerStats {
         version: string;
-    
+
```
There is a jslint rule to avoid these issues: https://eslint.org/docs/rules/no-trailing-spaces. Please enable it in this file: https://github.com/Ericsson/CodecheckerVSCodePlugin/blob/main/.eslintrc.json. So we can avoid these problems next time.

Discookie replied:
I'll add this rule (along with a couple of others) in a separate PR, after this one is merged.

csordasmarton replied:
Okay, let's do it in a separate PR but I have created an issue for it (#11).

csordasmarton commented on src/sidebar/views/reports.ts:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    getChildren(element?: IssueMetadata): IssueMetadata[] | undefined {
+        const makeArray = <T>(length: number, func: (idx: number) => T): T[] => {
```
This variable / function is not used so it can be removed. To avoid such problems please enable this rule: https://eslint.org/docs/rules/no-unused-vars

Discookie commented on src/sidebar/views/reports.ts:
```diff
@@ -0,0 +1,249 @@
+import { basename } from 'path';
```
The 'Related files' section displays errors in headers and such, but they are in the currently opened file's ``.plist`` file.

To keep the errors when changing files, either:
* the source-file needs to be tracked (via eg. DiagnosticsAPI.activeReport), which only provides one previous files' error,
* all the reports need to be pre-processed, to see which report contains errors for the file,
* or the found diagnostics need to be sorted at runtime.

Originally I implemented the first option, and it resulted in very inconsistent behavior.
The third option shouldn't be too difficult to implement, but it isn't straightforward either, and the results either need to be cached, or the displayed errors will be inconsistent when the user restarts their editor.

csordasmarton replied:
My real problem is that we are hiding bugs when we don't show reports from header files. So if someone modifies only a  header file but introduces a problem on re-analysis the bug will not be shown on the 'Reports in current file' view.

We don't have to solve this problem in this PR, but I created an issue for this problem (#10) to not forget it.

Discookie commented on CHANGELOG.md:
```diff
@@ -10,4 +10,5 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
 - Parser and processing logic for ``metadata.json``
-- Sidebar tab, and an Overview view with basic stats
+- Parser and processing logic for ``.plist`` files
+- Sidebar tab, including an Overview section with basic stats, and an Issues section for the current file's issues
```
Updated the changelog with a couple of entries.

Discookie commented:
Added the final part of this PR - navigation.

* The ``Jump to report`` button now works, as well as clicking on each reproduction path member.
* A button for displaying the repr. path is added.
  * Hotkey navigation in the path is possible with Ctrl/Ctrl-Shift-F7.
* Added a couple of tests for the parser, but not for the processor.

Currently only the repr. path is kept when switching between files, which comes with a couple of inconsistencies of the displayed reports. Those will be addressed by #10, and they're out of scope for this PR.

And with that, this PR is ready for review and merge.

csordasmarton commented on src/backend/processor/diagnostics.ts:
```diff
@@ -0,0 +1,204 @@
+
+    // Logs errors to the user, and always resolves.
+    async reloadDiagnosticsSafe(forceReload?: boolean): Promise<void> {
+        try {
```
The regular try/catch is not able to catch exception from an async function without await:
```ts
async function foo(): Promise<void> {
  return new Promise((res, rej) => {
    rej("Failure");
  })
}

async function fooSafe(): Promise<void> {
  try {
    foo();
  } catch (err) {
    console.error("fooSafe", err);
  }
}

async function fooSafe2(): Promise<void> {
  try {
    await foo();
  } catch (err) {
    console.error("fooSafe2", err);
  }
}

async function fooSafe3(): Promise<void> {
  foo().then(() => {
    console.log("SUCCESS");
  }).catch(err => {
    console.error("fooSafe3", err);
  });
}

// This will not catch the error.
fooSafe();

// Using await. This will catch the error.
fooSafe2();

// This will catch the error too.
fooSafe3();
```

Discookie replied:
Thank you, fixed.

csordasmarton commented on src/editor/navigation.ts:
```diff
@@ -0,0 +1,190 @@
+        const entry = ExtensionApi.diagnostics.selectedEntry?.diagnostic;
+
+        if (stepIdx === null || !entry) {
+            return;
```
- If I open a file and the Codechecker view but I do not select any report and I hit ctrl+f7 this function will be called but entry will be null and we just simply return from this function but I didn't get any message /popup that says I have to select a report to jump to the next step.
- Also if I open a bug / clicking to the `Jump to report` button or to a bug step (selecting a report / bug step) and using ctrl+f7, it will do the same thing, so nothing.
- If I select a report and clicking to the `Show reproduction steps` button and then using the ctrl+f7 key combination the feature works.

Discookie replied:
This is intentional, much like how you cannot `Jump to next error in file` if there are no errors in the file... :D
The 'first/last reproduction step' display is also hidden if the path is not shown.

I can set a condition for the hotkey, but this early return leads to a much simpler logic.

csordasmarton replied:
If I don't select a bug step, it is not annoying if this feature do nothing. But in the second case if I select the first bug step and the report has multiple bug step I expect that this feature should work.

Discookie replied:
Fixed so the repr. path is shown when the user jumps to a step.
It is very unintuitive, though, because I don't have a simple logic for automatically removing a shown path.
It works for now, but the logic will need a rework as part of #10.

csordasmarton reviewed (approved):
LGTM!
